File: algo_2021/w4.-queue-bfs/problems/457.Implement_Queue_by_Linked_List/457_implement_queue_by_linked_list__linkedlist.java
class Node {
   public int val;
   public Node next;
   public Node(int _val) {
       val = _val;
       next = null;
   } 
}

public class MyQueue {
    /*
     * @param item: An integer
     * @return: nothing
     */
    public Node head, tail;
    public void enqueue(int item) {
        // write your code here
        if (head == null) {
            head = new Node(item);
            tail = head;
        } else {
            tail.next = new Node(item);
            tail = tail.next;
        }  
    }

    /*
     * @return: An integer
     */
    public int dequeue() {
        // write your code here
        if (head == null) {
            return -1000;
        }
        int item = head.val;
        head = head.next;
        return item;
    }
}

File: lintcode_algo_2021/s4.-queue-bfs-graph/457.Implement_Queue_by_Linked_List/457_implement_queue_by_linked_list__linkedlist.py
class Node:
    def __init__(self, _val):
        self.val = _val
        self.next = None
        

class MyQueue:
    """
    @param: item: An integer
    @return: nothing
    """
    def __init__(self):
        self.head, self.tail = None, None

    def enqueue(self, item):
        # write your code here
        if not self.head:
            self.head = Node(item)
            self.tail = self.head
        else:
            self.tail.next = Node(item)
            self.tail = self.tail.next

    """
    @return: An integer
    """
    def dequeue(self):
        # write your code here
        # queue empty
        if not self.head:
            return -1000
        item = self.head.val
        self.head = self.head.next
        return item
